provisioner/terraform/timings.go

```diff
@@ -0,0 +1,238 @@
+	timingInitStart     timingKind = "init_start"
+	timingInitComplete  timingKind = "init_complete"
+	timingInitErrored   timingKind = "init_errored"
+	timingGraphStart    timingKind = "graph_start"
+	timingGraphComplete timingKind = "graph_complete"
+	timingGraphErrored  timingKind = "graph_errored"
+	// Other terraform log types which we ignore.
+	timingLog        timingKind = "log"
+	timingInitOutput timingKind = "init_output"
+)
+
+type timingAggregator struct {
+	mu sync.Mutex
+
+	stage       database.ProvisionerJobTimingStage
+	stateLookup map[uint64]*timingSpan
+}
+
+type timingSpan struct {
+	kind                       timingKind
+	start, end                 time.Time
+	stage                      database.ProvisionerJobTimingStage
+	action, provider, resource string
+	state                      proto.TimingState
+}
+
+// newTimingAggregator creates a new aggregator which measures the duration of resource init/plan/apply actions; stage
+// represents the stage of provisioning the timings are occurring within.
+func newTimingAggregator(stage database.ProvisionerJobTimingStage) *timingAggregator {
+	return &timingAggregator{
+		stage:       stage,
+		stateLookup: make(map[uint64]*timingSpan),
+	}
+}
+
+// ingest accepts a timing span at a certain timestamp and assigns it a state according to the kind of timing event.
+// We memoize start & completion events, and then calculate their total duration in aggregate.
+// We ignore progress events because we only care about the full duration of the action (delta between *_start and *_complete events).
+func (t *timingAggregator) ingest(ts time.Time, s *timingSpan) {
+	if s == nil {
+		return
+	}
+
+	s.stage = t.stage
+	ts = dbtime.Time(ts)
+
+	switch s.kind {
+	case timingApplyStart, timingProvisionStart, timingRefreshStart, timingInitStart, timingGraphStart:
+		s.start = ts
+		s.state = proto.TimingState_STARTED
+	case timingApplyComplete, timingProvisionComplete, timingRefreshComplete, timingInitComplete, timingGraphComplete:
+		s.end = ts
+		s.state = proto.TimingState_COMPLETED
+	case timingApplyErrored, timingProvisionErrored, timingInitErrored, timingGraphErrored:
+		s.end = ts
+		s.state = proto.TimingState_FAILED
+	default:
+		// We just want start/end timings, ignore all other events.
+		return
+	}
+
+	t.mu.Lock()
+	// Memoize this span by its unique attributes and the determined state.
+	// This will be used in aggregate() to determine the duration of the resource action.
+	t.stateLookup[s.hashByState(s.state)] = s
+	t.mu.Unlock()
+}
+
+// aggregate performs a pass through all memoized events to build up a slice of *proto.Timing instances which represent
+// the total time taken to perform a certain action.
+// The resulting slice of *proto.Timing is NOT sorted.
+func (t *timingAggregator) aggregate() []*proto.Timing {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	out := make([]*proto.Timing, 0, len(t.stateLookup))
```

Comment:
I feel like this could be simplified, depending on how much we end up skipping the minor allocation saving is probably a wash?
```suggestion
	out := []*proto.Timing{}
```

Reply:
Better to have ~loved and lost than never to have loved at all~ allocated a small bit of memory upfront then potentially incur multiple allocs later. IIRC slices start with a cap of 4 then double each time it's filled; if we have 100 items to manage that's ~5 allocs instead of 1, and the process of allocing is what's really slow AFAIK.

Reply:
I'll leave it up to you.

That said, I don't think pre-allocating the slice will change allocs significantly considering we're already using a slice of pointers. I wrote a quick benchmark, and it may be a poor comparison, but this is the result:

```
❯ go test -bench .
goos: darwin
goarch: arm64
BenchmarkAlloc-10           	  750336	      1579 ns/op	    2976 B/op	     109 allocs/op
BenchmarkAllocUpfront-10    	  972052	      1210 ns/op	    1704 B/op	     102 allocs/op
PASS
```

https://go.dev/play/p/sQOsCPL4fW9